cranelift-codegen/src/regalloc/spilling.rs

```diff
@@ -395,6 +433,11 @@ impl<'a> Context<'a> {
             if abi.location.is_reg() {
                 let (rci, spilled) = match self.liveness[arg].affinity {
                     Affinity::Reg(rci) => (rci, false),
+                    Affinity::RegUnit(_unit) => (
```

Comment:
RegUse tracks at register class granularity, so I'm following that and assuming that the value's affinity will not conflict with where the ABI says its value type should be. That seems like a safe assumption, but I wanted to call it out here in case it's not

Reply:
It's not immediately obvious that it's safe. See `Affinity::merge` for example where it makes no change if there are conflicting register class constraints.

Reply:
Wouldn't a conflict between the constraints here be a logical error? A value's affinity shouldn't place it in a register class that can't hold those values. (I realize since then that the `Reg` case is similar -- uses the declared register affinity regardless of `abi.value_type`. The equivalent here would be `self.reginfo.toprc_for_reguint(unit)`)

Reply:
The ABI might request the value in one sub-register class, and the affinity might be for a different sub-register class. They'd both be part of the same parent register class, which is valid for the type, but they wouldn't intersect. Would that cause a problem here?

Reply:
I see, that's a good question. Here's my understanding, which leads me to think this can be resolved without issue:

`collect_abi_reg_uses` collects uses of registers for argument values as they exist when reaching the call, so the register class reported here should be derived from the existing affinity (and my arm for `RegUnit` is incorrect). This explains why `RegClass` is correct in returning `rci`, even if that disagrees with the ABI-requested location. `process_reg_uses` then inserts copies, but I don't entirely understand that logic. I _think_ that by widening back to a register class that contains `unit` in this arm, the worst that could happen is that we decide to copy `Value`, ending its live range with an `Affinity::RegClass(rci)`, which might be wider than the specific affinity the use came from. That would force some extra copies, but I don't think would be incorrect. The value's copy wouldn't have any set location, and would later be colored into the ABI-correct register. In this case, we can come back later and make `RegUse` smarter by making it aware of specific register affinities, and avoiding the `insert_copy` if it would copy to where it needs to go anyway?

I think this all only applies if we need to reload for call arguments, and I'm not really sure how I'd construct a test case for this. Either way, I'll change this to return a register class derived from `unit`, and see if that still works - I'd hope that trying to end the value's live range with an affinity that causes a conflict would at least cause a panic in codegen, so not hitting that even with some fuzzing is a bit perplexing to me.. It could be that for x86 the bug case isn't reachable here; `regclass_for_abi_type` would give GPR, the same as `toprc_for(unit)`, and wouldn't be a conflict.

If the above reads correct to you, I'd like to expand on `collect_abi_reg_uses`'s docs a bit to clarify how it cares about the values' affinities - namely that it's interested in where they currently are, not where the function will eventually want them, and that the function's ABI at this point is not important here.